File: 2020/python/day02.py
import re

#####################################################################

def main():
  lines = read_input()
  print(f'part 1: {part_1(lines)}')
  print(f'part 2: {part_2(lines)}')

def part_1(lines):
  return sum(is_valid_1(line) for line in lines)

def part_2(lines):
  return sum(is_valid_2(line) for line in lines)

#####################################################################

def read_input():
  def parse_line(line):
    values = re.match('^(\d+)-(\d+) (\w): (\w+)$', line).groups()
    return (int(values[0]), int(values[1]), values[2], values[3])
  with open('../inputs/02.txt') as f:
    return [parse_line(line) for line in f.readlines()]

def is_valid_1(line):
  low, high, letter, password = line
  letter_count = password.count(letter) 
  return low <= letter_count <= high

def is_valid_2(line):
  low, high, letter, password = line
  return (password[low-1] == letter) ^ (password[high-1] == letter)

#####################################################################

if __name__ == "__main__":
  main()

File: 2020/haskell/Day02.hs
module Day02 where

import Text.Printf(printf)

-- Main zone--
main = do
  input <- readFile "../inputs/02.txt"
  printf "part 1: %d\n" $ part1 input
  printf "part 2: %d\n" $ part2 input

part1 input = length $ filter isValid1 $ makeLines input
part2 input = length $ filter isValid2 $ makeLines input

-- For part 1 --
count :: Char -> String -> Int
count c s = length $ filter (c==) s

isValid1 :: Line -> Bool
isValid1 (low, high, letter, password) = 
  letter_count >= low && letter_count <= high
  where letter_count = count letter password

-- For part 2 --
isValid2 :: Line -> Bool
isValid2 (low, high, letter, password) = 
  (password!!(low-1) == letter) /= (password!!(high-1) == letter) 

-- For both parts --
type Line = (Int, Int, Char, String)

replace :: Char -> Char -> String -> String
replace c1 c2 s = map (\c -> if c==c1 then c2 else c) s

parse :: String -> [String]
parse s = words $ replace '-' ' ' (filter (':'/=) s)

makeLine :: [String] -> Line
makeLine (a:b:c:d:_) = (read a, read b, head c, d)

makeLines :: String -> [Line]
makeLines input = map (makeLine . parse) $ lines input
